### ISSUE TITLE
Is this a BUG in blob->offset(n) ?

### ISSUE THREAD

**Issue by Coldmooon** [BVLC/caffe#3420]
Say the input shape is 10 * 192 * 32 * 32.
Then:
bottom[0]->offset(0) will be 0.
bottom[0]->offset(1) will be 196608.
bottom[0]->offset(2) will be 393216.
...
bottom[0]->offset(9) will be 1769472.
bottom[0]->offset(10) will be 1966080.
There isn't bottom[0]->offset(11) .

The last one in a batch is in offset(9) because we start at zero. So,  I think there shouldn't be offset(10). 
If we read bottom[0]->cpu_data()[1966080], the code will exit without any error prompt.

**Comment by seanbell** [comment]
Yes, this is a bug and it should be `CHECK_LT(n, num())`.  However, in order to fix this, you will need to also fix a few other places that depend on the current behavior of `offset`.  While this bug has been pointed out many times, this is the reason why it's not yet fixed.  See #792, #1681, #2835.

**Comment by Coldmooon** [comment]
@seanbell This is a big project and needs programmers to well understand how layers work. I just found this bug in "prelu_layer.cu". By the way, could you tell me how you track issues and quickly know what issues are related to the current issue and should be referred? Because #792, #1681, #2835 you listed are very helpful to me. Thank you ~~

**Comment by seanbell** [comment]
I just searched the project for issues, using the builtin GitHub search.  I forget what search terms I used.

**Comment by Coldmooon** [comment]
@seanbell Very helpful~Thanks~

**Comment by Noiredd** [comment]
There are two open PRs which aim to fix this: #2835 and #6021. There was also #792 which IMHO was better as it also checked `>=0` (currently the check is placed on the blob dimensions instead of arguments, which I think is a mistake).  
Both open PRs contain an identical patch, which does not pass the tests. Turns out that other code depends on this broken behaviour and the change is not as simple as that.

**Comment by Chessing234** [comment]
Yeah, CHECK_LE(n, num()) should be CHECK_LT — same for c/h/w. Opened a PR for it.
